client.go

```diff
@@ -321,17 +322,28 @@ func (c *Client) Walk(root string) *fs.Walker {
 	return fs.WalkFS(root, c)
 }
 
-// ReadDir reads the directory named by dirname and returns a list of
-// directory entries.
+// ReadDir reads the directory named by p
+// and returns a list of directory entries.
 func (c *Client) ReadDir(p string) ([]os.FileInfo, error) {
+	return c.ReadDirContext(context.Background(), p)
+}
+
+// ReadDirContext reads the directory named by p
+// and returns a list of directory entries.
+// The passed context can be used to cancel the operation
+// returning all entries listed up to the cancellation.
+func (c *Client) ReadDirContext(ctx context.Context, p string) ([]os.FileInfo, error) {
 	handle, err := c.opendir(p)
 	if err != nil {
 		return nil, err
 	}
 	defer c.close(handle) // this has to defer earlier than the lock below
-	var attrs []os.FileInfo
+	var entries []os.FileInfo
 	var done = false
 	for !done {
+		if err = ctx.Err(); err != nil {
+			return entries, err
+		}
 		id := c.nextID()
 		typ, data, err1 := c.sendPacket(nil, &sshFxpReaddirPacket{
```

Comment:
This creates internally a channel, that is used to waiting on for a response from the server. https://github.com/pkg/sftp/blob/master/conn.go#L131

Instead we would want to use `dispatchRequest` and then we can:
```
select {
case <-ctx.Done():
  return entries, err
case result := <-ch:
  typ, data, err1 = s.typ, s.data, s.err
}
```

Otherwise, we’re going to be stuck until the server responds with at least one `sshFxpName` packet.

Reply:
Right, I have added the context to `opendir` and `sendPacket`